src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php

```diff
@@ -150,6 +150,38 @@ public function createListFromLoader(ChoiceLoaderInterface $loader, $value = nul
     }
 
     /**
+     * Filters a loaded choice list.
+     *
+     * The callable should be compatible with array_filter and
+     * return true to keep a loaded choice or false to discard it.
+     *
+     * @param ChoiceListInterface $choiceList The loaded choice list to filter
+     * @param callable            $filter     The filter to apply to the choices
+     *
+     * @return ChoiceListInterface The filtered choice list
+     */
+    public function filterChoiceList(ChoiceListInterface $choiceList, callable $filter)
```

Comment:
btw, I think this logic should not be in the factory at all. It is a totally different business logic than creating choice lists, which becomes clear when removing duplication in your code:

```php
$choices = array_filter($choiceList->getChoices(), $filter);

return $this->createListFromChoices($choices);
```

This means that the filtering should be moved to a consumer of the factory, which would then call the public API of the factory (supporting any factory, not only the caching one)